### ISSUE TITLE
Angular2 routing - broken back button in OS X Safari - Version 9.1 (11601.5.17.1)

### ISSUE THREAD

**Issue by hamzatayeb** [angular/angular#7873]
The demo uses a simple Quiz application (just the routing part).

Steps to reproduce -
1. Start the quiz
2. Navigate forward
3. Click the back button and notice that the component does not render properly

See Plunker - http://plnkr.co/edit/8LMo0VLCLIuM06itCifn?p=preview

**Current behavior**
- usage of the browser's back button does not cause component to re-render properly

**Expected/desired behavior**
- usage of the browser's back button *does* cause the component to re-render properly (see Google Chrome behavior for expected behavior)

**Other information**
I also posted a [question to Stack Overflow](http://stackoverflow.com/questions/36361140/back-button-on-angular2-is-broken) about this for additional context

**Comment by d8660091** [comment]
I have the same problem. In fact, it also happens to the example given by the official website (https://angular.io/resources/live-examples/toh-5/ts/plnkr.html) when using safari.

**Comment by MartinMa** [comment]
I think the problem is that `OnInit` is not called, when using the back button. This issue might be a duplicate of #7722.

**Comment by goodfield** [comment]
Got the same problem in Safari, IE11 and IE10. 

When using the back/forward button or when i manipulate the url ngAfterContentInit and ngAfterViewInit don't get called.

I did some analysis, looks like ngAfterContentInit and ngAfterViewInit get called when tick() get's callen in the onMicrotaskEmpty event:

```typescript
export class ApplicationRef_ extends ApplicationRef {
constructor(private _platform: PlatformRef_, private _zone: NgZone, private _injector: Injector) {
    super();
    if (isPresent(this._zone)) {
      ObservableWrapper.subscribe(this._zone.onMicrotaskEmpty,
                                  (_) => { this._zone.run(() => { this.tick(); }); });
    }
    this._enforceNoNewChanges = assertionsEnabled();
  }
```

On Safari, IE11 and IE10 this never happens when using Back/Forward Button.

As really dirty workaround (not sure if it has any other impacts) i did following:

```typescript
import {Component, ApplicationRef} from 'angular2/core';
import {RouteConfig, ROUTER_DIRECTIVES, Router} from 'angular2/router';

import {AComponent} from './a.component';
import {BComponent} from './b.component';

@Component({
  selector: 'my-app',
  template: '<h1>My First Angular 2 App</h1><router-outlet></router-outlet>',
  directives: [ROUTER_DIRECTIVES]
})
@RouteConfig([
  {path: '/a', name: 'A', component: AComponent, useAsDefault: true},
  {path: '/b', name: 'B', component: BComponent}
])
export class AppComponent {

  constructor(private _applicationRef:ApplicationRef, private _router:Router) {
    _router.subscribe(() => {
      this._applicationRef.tick();
      setTimeout(() => {
        this._applicationRef.tick();
      }, 100);
    });

  }
}
```

**Comment by ravinderpayal** [comment]
Same problem here 
[StackOverflow Question](http://stackoverflow.com/questions/37192628/history-forward-backward-button-doesnt-work-angular-2-router)

**Comment by gillestasse** [comment]
I am a similar issue using Chrome. The back button only works once. If i navigate :

1. Page 1
1. Page 2
1. Page 3

I can only go back to Page 2, pressing the back button once more does not return me to Page 1.

**Comment by feliu** [comment]
+1 same problem..

**Comment by zoechi** [comment]
+1 is deprecated. Use "Add reaction" instead https://github.com/blog/2119-add-reactions-to-pull-requests-issues-and-comments

Rather provide a Plunker that allows to reproduce the problem.

**Comment by ianbelcher** [comment]
I struggled with this for a bit today but I eventually found a temporary solution listening for changes on the router. Its only a temporary fix but at least you can get Safari working again!
[StackOverflow description](http://stackoverflow.com/a/37521006/981598)

**Comment by ianbelcher** [comment]
Adding the check `/*@cc_on!@*/false || !!document.documentMode` to the above `if` statement will also fix IE as well.

**Comment by daslicht** [comment]
is there meanwhile a suggested solution for this behavior ?, please ?

**Comment by justme1** [comment]
Any updates on this?

**Comment by ChrisMeeusen** [comment]
Thanks for posting the workaround.  I modified the constructor slightly to account for the new router.  Here it is if anyone is looking:

```
export class AppComponent {
    //temp work around for IE back button issue https://github.com/angular/angular/issues/7873
    constructor(private _applicationRef: ApplicationRef, private _router: `Router)` {
        
        _router.events.subscribe(() => {
            this._applicationRef.tick();
            setTimeout(() => {
                this._applicationRef.tick();
            }, 100);
        });

    }
```
the new router didn't have a subscribe method on it.  Is this going to be addressed in RC5?

**Comment by MartinMa** [comment]
I think this issue was fixed with this commit: f08060b0b00c98341c7e1fd9acb984402c280396
The new new new router has it's own release cycle. The fix will probably be part of the next release. Currently at `3.0.0-beta.2`.

**Comment by danduh** [comment]
Minor update on @goodfield 's solution:


```
export class AppComponent {
    privet subscription : Subscription;
     constructor(private _applicationRef: ApplicationRef, private _router: Router) {
         _router.events.subscribe(() => {
             this._applicationRef.tick();
             setTimeout(() => {
                 this._applicationRef.tick();
             }, 100);
         });
     }
  ngOnDestroy(){
     this.subscription.unsubscribe()
 }
```

**Comment by smith64fx** [comment]
I think this is fixed in RC6. Can you test again?

**Comment by smith64fx** [comment]
I think this is fixed in RC6. Can you test again?

**Comment by MartinMa** [comment]
This issue was alrady fixed with router `3.0.0-rc.1`. See [changelog](https://github.com/angular/angular/blob/master/modules/%40angular/router/CHANGELOG.md). Please close @vsavkin.

**Comment by AyzPat** [comment]
When i click the back button in my ng2 app the view is correct but the url is not correct according to the view. But the same works fine on click to the nav links...How to overcome it?
